File: C++/LinkedList.cpp
#include <bits/stdc++.h>
#include <conio.h>

using namespace std;

class Node
{
public:
    int data;
    Node *next;
    Node(int val)
    {
        this->data = val;
        this->next = NULL;
    }
};

class LinkedList
{

    Node *head = NULL;

public:
    void insert(int data)
    {
        Node *new_node = new Node(data);
        if (head == NULL)
        {
            head = new_node;
            return;
        }

        Node *temp = head;
        while (temp->next != NULL)
        {
            temp = temp->next;
        }
        temp->next = new_node;
    }

    void print()
    {
        Node *temp = head;

        while (temp != NULL)
        {
            cout << temp->data << "->";
            temp = temp->next;
        }
        cout << "NULL"
             << "\n";
    }

    void remove()
    {
        Node *temp = head;
        while (temp->next->next != NULL)
        {
            temp = temp->next;
        }
        temp->next = NULL;
    }

    void reverse()
    {
        Node *temp = head;
        Node *prev = NULL;
        while (temp != NULL)
        {
            Node *next = temp->next;
            temp->next = prev;
            prev = temp;
            temp = next;
        }

        head = prev;
    }
};

int main()
{
    LinkedList list;
    list.insert(4);
    list.insert(5);
    list.insert(6);
    list.insert(7);
    list.print();
    list.reverse();
    list.print();
    // list.remove();
    // list.print();
    return 0;
}

File: TypeScript/LinkedList.ts
 class ListNode {
      val: number
      next: ListNode | null
      constructor(val?: number, next?: ListNode | null) {
          this.val = (val===undefined ? 0 : val)
          this.next = (next===undefined ? null : next)
      }
  }


function reverseList(head: ListNode | null): ListNode | null {
      let temp = head
      let prev:ListNode|null = null
      
      while(temp != null){
          let next = temp.next
          temp.next = prev
          prev = temp
          temp = next
      }
    
    head = prev
     return head
};

function middleNode(head: ListNode | null): ListNode | null {
      let slow:ListNode|null = head;
      let fast:ListNode|null = head;
    
      while(fast && fast.next){
         slow = slow?.next
         fast = fast.next.next
      }
    return slow
};

function deleteNode(node: ListNode | null): void {
        let temp:ListNode | null = node
        temp?.val = temp?.next?.val
        temp?.next = temp?.next?.next
};